collector/task_reference_wait_time.go

```diff
@@ -0,0 +1,152 @@
+package collector
+
+import (
+	"context"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	"knative.dev/pkg/apis"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/metrics"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sync"
+)
+
+func SetupTaskReferenceWaitTimeController(mgr ctrl.Manager) error {
+	reconciler := &ReconcileTaskReferenceWaitTime{
+		client:        mgr.GetClient(),
+		scheme:        mgr.GetScheme(),
+		eventRecorder: mgr.GetEventRecorderFor("MetricExporterWaitTaskRunTaskResolution"),
+	}
+	waitMetric := NewTaskReferenceWaitTimeMetric()
+	metrics.Registry.MustRegister(waitMetric)
+	return ctrl.NewControllerManagedBy(mgr).For(&v1beta1.TaskRun{}).WithEventFilter(&taskRefWaitTimeFilter{waitDuration: waitMetric, transitionTime: map[string]metav1.Time{}, lock: sync.Mutex{}}).Complete(reconciler)
+}
+
+func NewTaskReferenceWaitTimeMetric() *prometheus.HistogramVec {
+	labelNames := []string{NS_LABEL, TASK_NAME_LABEL}
+	return prometheus.NewHistogramVec(prometheus.HistogramOpts{
+		Name:    "taskrun_task_resolution_wait_milliseconds",
+		Help:    "Duration in milliseconds for a resolution request for a task reference needed by a taskrun to be recognized as complete by the taskrun reconciler in the tekton controller. ",
+		Buckets: prometheus.ExponentialBuckets(float64(100), float64(5), 6),
+	}, labelNames)
+
+}
+
+type ReconcileTaskReferenceWaitTime struct {
+	client        client.Client
+	scheme        *runtime.Scheme
+	eventRecorder record.EventRecorder
+}
+
+type taskRefWaitTimeFilter struct {
+	waitDuration *prometheus.HistogramVec
+	// so knative/tekton allows for updates to a conditions last transition time without changing the reason of the condition,
```

Comment:
this is incorrect .... I found the knative logic that prevents this

I'll update this PR later today to remove the map/lock, and simplify